controllers/managedocs_controller_test.go

```diff
@@ -2,639 +2,759 @@ package controllers
+
+				// Updating the status of the prometheus resource should trigger a reconcile for managedocs
+				am.Status.UnavailableReplicas = 1
+				Expect(k8sClient.Update(ctx, am)).Should(Succeed())
+
+				managedOCS := managedOCSTemplate.DeepCopy()
+				key := utils.GetResourceKey(managedOCS)
+				Eventually(func() v1.ComponentState {
+					Expect(k8sClient.Get(ctx, key, managedOCS)).Should(Succeed())
+					return managedOCS.Status.Components.Alertmanager.State
+				}, timeout, interval).Should(Equal(v1.ComponentPending))
+			})
+		})
+		When("all alertmanager replicas are available", func() {
+			It("should reflect that in the ManagedOCS resource status", func() {
+				By("by seeting Status.Components.Alertmanager.State to Ready")
+				// Ensure that the prometheus resource has unavailable replicas
+				am := amTemplate.DeepCopy()
+				Expect(k8sClient.Get(ctx, utils.GetResourceKey(am), am)).Should(Succeed())
+
+				// Updating the status of the prometheus resource should trigger a reconcile for managedocs
+				am.Status.UnavailableReplicas = 0
+				Expect(k8sClient.Update(ctx, am)).Should(Succeed())
+
+				managedOCS := managedOCSTemplate.DeepCopy()
+				key := utils.GetResourceKey(managedOCS)
+				Eventually(func() v1.ComponentState {
+					Expect(k8sClient.Get(ctx, key, managedOCS)).Should(Succeed())
+					return managedOCS.Status.Components.Alertmanager.State
+				}, timeout, interval).Should(Equal(v1.ComponentPending))
+			})
+		})
+		When("the storagecluster resource is modified while the reconcile strategy is not set", func() {
+			It("should revert the changes and bring the resource back to its managed state", func() {
+				// Set managed OCS to reconcile strategy to strict
+				managedOCS := managedOCSTemplate.DeepCopy()
 				Expect(k8sClient.Get(ctx, utils.GetResourceKey(managedOCS), managedOCS)).Should(Succeed())
-				Expect(managedOCS.Status.ReconcileStrategy == v1.ReconcileStrategyStrict).Should(BeTrue())
+				managedOCS.Spec.ReconcileStrategy = ""
+				Expect(k8sClient.Update(ctx, managedOCS)).Should(Succeed())
 
-				sc := &ocsv1.StorageCluster{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      storageClusterName,
-						Namespace: TestNamespace,
-					},
-				}
+				// Get an updated storagecluster
+				sc := scTemplate.DeepCopy()
 				scKey := utils.GetResourceKey(sc)
 				Expect(k8sClient.Get(ctx, scKey, sc)).Should(Succeed())
 
-				// Modify the storagecluster spec
+				// Update to empty sepc
 				spec := sc.Spec.DeepCopy()
 				sc.Spec = ocsv1.StorageClusterSpec{}
 				Expect(k8sClient.Update(ctx, sc)).Should(Succeed())
 
-				// Wait for the storage cluster to be modfied again to reflect that it was reconciled
-				utils.WaitForResourceSpecToUpdate(k8sClient, ctx, sc, timeout, interval)
-
-				// Verify that the storage cluster was reverted to its original state
-				Expect(reflect.DeepEqual(sc.Spec, *spec)).Should(BeTrue())
+				// Wait for the spec changes to be reverted
```

Comment:
```suggestion
				// Reverting secret to reflect the current size.
				secret := addonParamsSecretTemplate.DeepCopy()                                  
                               secret.Data["size"] = []byte("4")
                               Expect(k8sClient.Update(ctx, secret)).Should(Succeed())                                                           

				// Wait for the spec changes to be reverted
```
This test is failing because of the downscaling test above. The secret is set to 1, so when the StorageCluster spec is modified, the deployer reconciles it, but the StorageCluster is made with the Count listed in the secret, since it doesn't see that the current count is higher. This can be fixed by updating the secret.

Reply:
I think this actually caught a fundamental problem we have. If we lose the size info on the storage cluster we cannot revert back to its actual state only to the desired state. I can reorder the tests to prevent that from happening, but to solve the underlying problem we will need to think about what does it mean to revert back and what should be the state to revert to.

@renan-campos  I suggest to merge with this single failing test, and let it fail until we have a proper solution

Reply:
Well put. I think an issue should be made here in the repo to capture the context to the problem that you described so it is easily found when it gets addressed.

Do we really want to have a failing test in the codebase? This doesn't need to be immediately addressed, and the suggested change reflects the current workaround: we rely on the secret to keep track of the size. I prefer to have an issue listed in this repo and comments linking it in a passing test over having to remember that there is one test which will fail when I pull and run the tests.

Reply:
A failing test describes the result we would like to achieve and it is great as a constant reminder that we need to fix it. Saying that I agree that we should open a github issue to capture the Bug.

Reply:
If we do not want to fail the test we need to mark it as a skipped test in the result (not sure how to do it)

Reply:
For now, I decided to reorder the tests in a way that they will not fail